### PR TITLE
Add artifactory clean

### PR DESCRIPTION
**Pull request by boxboatmatt**
Signed-off-by: Matthew DeVenny <matt@boxboat.com>

### CODE REVIEW

**Review by caleblloyd — CHANGES_REQUESTED** [review]

File: src/com/boxboat/jenkins/library/docker/ArtifactoryClean.groovy [hunk 1]
```diff
@@ -0,0 +1,127 @@
+package com.boxboat.jenkins.library.docker
+
+import com.boxboat.jenkins.library.config.Config
+
+class ArtifactoryClean implements Serializable {
+
+    boolean dryRun
+    int retentionDays
+    int removalCount = 0
+    List<String> registryKeys
```
[the remark this answers is not in the document]
Comment (boxboatmatt) [reply src/com/boxboat/jenkins/library/docker/ArtifactoryClean.groovy:10]: I don't think we can infer anything... In the current example usage `registryMap["default"].namespace` is `docker` but the `dockerRepo` is `docker-local`, `dockerRepoPath` is `apps/*`, and `registryKeys` refers to the registries defined in `config.yaml` `registryMap`. Perhaps there are better names?

File: src/com/boxboat/jenkins/pipeline/common/clean/ArtifactoryCleanHelper.groovy [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    static class ArtifactoryCleanParams extends BaseConfig<ArtifactoryCleanParams> implements Serializable {
+        boolean dryRun = false
+        int retentionDays = 15
+        List<ArtifactoryCleanRegistry> artifactoryCleanRegistries
+    }
+
+    static String clean(Map params) {
+        def paramsObj = new ArtifactoryCleanParams().newFromObject(params)
+        if (!paramsObj.artifactoryCleanRegistries) {
+            Config.pipeline.error "'registryKeys' is required"
```
Comment (caleblloyd) [inline src/com/boxboat/jenkins/pipeline/common/clean/ArtifactoryCleanHelper.groovy:19]: error should be `artifactoryCleanRegistries` not `registryKeys`

File: src/com/boxboat/jenkins/library/docker/ArtifactoryClean.groovy [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    boolean dryRun
+    int retentionDays
+    int removalCount = 0
+    List<ArtifactoryCleanRegistry> artifactoryCleanRegistries
+    String cleanSummary
+
+    Map<String, List<ArtifactoryImageManifest>> readRepositories(ArtifactoryCleanRegistry cleanRegistry) {
+        def registry = Config.global.getRegistry(cleanRegistry.registryKey)
+        def requestURI = registry.getRegistryUrl() + "/artifactory/api/search/aql"
+        def pathMatch = cleanRegistry.path == null ? "*" : cleanRegistry.path
```
Comment (caleblloyd) [inline src/com/boxboat/jenkins/library/docker/ArtifactoryClean.groovy:16]: nitpick - use the Elvis Operator - shorter and also matches empty string

```
def pathMatch = cleanRegistry.path ?: "*"
```

**Review by caleblloyd — CHANGES_REQUESTED** [review]

File: src/com/boxboat/jenkins/library/docker/ArtifactoryClean.groovy (hunk 3 above)
Comment (boxboatmatt) [reply src/com/boxboat/jenkins/library/docker/ArtifactoryClean.groovy:16]: think we should rename `cleanRegistry.path` to `cleanRegistry.pathMatch`?

Comment (caleblloyd) [reply src/com/boxboat/jenkins/library/docker/ArtifactoryClean.groovy:16]: that's probably a good idea

**Review by caleblloyd — APPROVED** [review]

**Pull request merged**